### ISSUE TITLE
`npm prune --production` issue

### ISSUE THREAD

**Issue by chenrui333** [lerna/lerna#2196]
## Current Behavior

Currently the package setup is like this:

```
├── package-a
├── package-b
└── package-core
```

both package-a and package-b depends on package-core (which is also an unpublished package).

We did run `npm prune --production` per each package, otherwise the zip file is too big for lambda. But it ended up deleting the local dependency (package-core).

## Expected Behavior

I wonder what would be the best way to use `npm prune --production`.

NOTE, I could not run `npm install --production` (as that local dependency `package-core` is managed thru lerna).

## Steps to Reproduce (for bugs)

1. lerna bootstrap
2. lerna exec -- npm run build (rough command)
3. lerna exec -- npm prune --production (rough command)

<details><summary>lerna.json</summary><p>

```
$ cat lerna.json
{
  "lerna": "3.16.1",
  "version": "independent",
  "packages": [
    "packages/*"
  ],
  "npmClient": "npm",
  "npmClientArgs": [
    "--no-lockfile"
  ],
  "useWorkspaces": false
}
```

```json
<!-- Please paste your `lerna.json` here -->
```
</p></details>

<details><summary>lerna-debug.log</summary><p>

```txt
<!-- If you have a `lerna-debug.log` available, please paste it here -->
<!-- Otherwise, feel free to delete this <details> block -->
```
</p></details>

## Your Environment

| Executable | Version |
| ---: | :--- |
| `lerna --version` | 3.16.1 |
| `npm --version`  | 6.4.1 |
| `yarn --version` | VERSION |
| `node --version` | v8.16.0 |

| OS | Version |
| --- | --- |
| macOS Mojave | 10.14.5 |

**Comment by evocateur** [comment]
If you use relative `file:` specifiers to associate the sibling dependencies, it turns out you _can_ run `npm install --production` in the leaf directories. Run `lerna link convert` to automatically convert your local packages.

That being said, Lerna is not a deployment tool. Use `npm install` to install your lerna-published packages and create a deployable artifact from that process, not `lerna bootstrap`. Check out [serverless-jetpack](https://github.com/FormidableLabs/serverless-jetpack) for a more appropriate solution.

**Comment by chenrui333** [comment]
yeah, lerna is not supposed to deal with deployment, but now the issue is our SAM artifact is too big, that is why I come back to rely on the `npm prune --production` to remove non-production dependencies. 

I am actually not using `file:`, instead, I am using `"@xx/package-core": "^1.0.0"` (1.0.0 is the same version number across root package.json as well sub-package.json).

Let me know if it makes more sense now.

**Comment by tristanHessell** [comment]
@chenrui333 Apologies to necropost, but in case you (or someone else) needs a potental solution.

**Comment by joshystuart** [comment]
I was able to run `lerna bootstrap -- --production` to remove dev dependencies. eg.

```json
"scripts": {
    "bootstrap": "npm run install && lerna bootstrap",
    "prune": "lerna bootstrap -- --production && npm run install --production",
    "build": "lerna run build"
}
```

Then within my `Dockerfile` I could:

```
npm run bootstrap && npm run build && npm run prune
```

**Comment by marcesengel** [comment]
When you had previously already installed your dependencies you can go for a `lerna clean` before doing `lerna bootstrap -- --production` - otherwise the extraneous dependencies will be kept.

**Comment by SebTVisage** [comment]
I ran into the same problem and simply re-did a `npx lerna link` after the prune. 

This seemed with node 12. In node 16, it throws a 404 error when trying to run `npm prune --production` saying it can't find my local package (it's not published)

**Comment by JamesHenry** [comment]
Hi Folks,

This issue related to the legacy package management capabilities of Lerna. The world in which Lerna had to manage package installation and linking concerns because of gaps in the ecosystem no longer exists, we are blessed to have many great options now, not just `npm` (which has also massively improved since the early days).

In June 2023, in Lerna `7.0.0`, we removed the `lerna bootstrap`, `lerna add` and `lerna link` commands in Lerna by default. Over two years later, in September 2025 (Lerna `9.0.0`), the commands were finally fully removed.

We provided some guidance at the time on how to migrate existing codebases to use native package manager features here: https://lerna.js.org/docs/legacy-package-management

I'll close this one as no action is applicable, but please always feel free to open a fresh issue giving as much context as possible when using the latest and greatest and we'll take a look. I'm making sure we close older issues like this one to free up bandwidth to make sure we tackle what matters most to users in 2026.

Many thanks for your contributions to Lerna!
